Translate part 3 from Russian to English
Add sections about distinct operator and interval function

Slightly modify several examples (#7)
this-is-learning/rxjs-fundamentals-course: part-3/part-3.md

Contributors:

- Andrew Grekov

> - Armen Vardanyan

Reviewer: Oooh yeeaaah 😃

Author: 😎

#

# **RxJS operators that we use on a daily basis**

# Comment

> When we only begin to work with **RxJS**, we need to learn about several operators that we will be using all the time. In this part of the course we will learn about the most common operators and also use them in practice.

Author: Sure thin

source.subscribe((value) => console.log(value));
```

We replaced **filter()** with **skip()**, which made our code more readable and understandable. Now our code looks more declarative.

Now let's do the opposite, that is, take only the first several values from a stream, and ignore the rest. For that purpose, we will need operators **take()** and **first()**.

**take()** takes the first several values in the stream (provided in the argument), and completes the stream.

> **first()** takes only the very first value from the stream and complete it. Notice that if the stream completes itself before emitting any value, an `EmptyError` will be thrown, so **first()** is fundamentally different from **take(1)**.

Reviewer: ```suggestion
**first()** takes only the very first value from the stream, then completes it. Notice that if the stream completes before emitting any value, an `EmptyError` will be thrown, so **first()** is fundamentally different from **take(1)**.
```

```ts
import { from } from 'rxjs';

import { take, first } from 'rxjs/operators';

const source = from([1, 2, 3]);

const sourceTakeTwo = source.pipe(

>   // will take the first two values from the stream and complete it 

Reviewer: ```suggestion
  // Will take the first two values from the stream and complete it 
```

  take(2),
);

// Will log:

//. 1

//. 2

sourceTakeTwo.subscribe(value => console.log(value));

const sourceTakeFirst = source.pipe(

>   // Will take the ver first value and complete the stream

Reviewer: ```suggestion
  // Will take the very first value and complete the stream
```

  first(),
);

// Will log:

//. 1

sourceTakeFirst.subscribe(value => console.log(value));
```

> Great!Now we know how to take the first values from a stream. Let's see how else can we filter values using **RxJS** operators.

Reviewer: ```suggestion
Great! Now that we know how to take the first values from a stream. Let's see how else can we filter values using **RxJS** operators.
```

Author: "Now that we know" implies there is going  to be a mention of what happens "now that we know", isn't it?

Reviewer: Correct, @Armenvardanyan95. How about this?
```suggestion
Great! Now that we know how to take the first values from a stream, let's see how else can we filter values using **RxJS** operators.
```

Author: Yep. also adding this

## **distinct()**

> Sometimes we need to only handle unique values, or in other words, ignore duplicates. For that purposes we can utilize the **distinct** operator. It will allow only those values that have not been emitted yet.

Reviewer: ```suggestion
Sometimes we only want to operate on unique values, or in other words, ignore duplicates. For that purposes we can utilize the **distinct** operator. It will only pass values that have not been emitted yet.
```

import { distinct } from 'rxjs/operators';

const source = from([1, 2, 2, 1, 3, 2, 3, 1, 3, 2, 1, 1, 3]).pipe(
  // Will skip duplicates

  distinct(),
);

// Will log:

//. 1

//. 2

//. 3

source.subscribe(value => console.log(value));
```

## **distinctUntilChanged()**

Sometimes though, we will need to skip duplicates, but only if they come one after another. For that, we can use the  **distinctUntilChanged()** operator.

**distinctUntilChanged()** skips duplicate values that immediately follow each other.

```ts
import { from } from 'rxjs';

import { distinctUntilChanged } from 'rxjs/operators';

const source = from([1, 1, 2, 2, 2, 3, 3, 1, 1]).pipe(
  // Will skip the value, if it is a duplicate of the previous one

  distinctUntilChanged(),
);

// Will log:

//. 1

//. 2

//. 3

//. 1

source.subscribe(value => console.log(value));
```

As we can see, **distinctUntilChanged()** skipped only the first duplicates.

#

# **Combining streams**

Sometimes we deal with tasks that require us to handle several streams simultaneously. For starters, let's create some streams to work with. For that purpose we will use the  **timer()** function.

## **timer()**

> **timer()** creates a stream that emits a value the time (in milliseconds) we provide has passed, essentially working like `setTimeout`, but for streams.

Reviewer: ```suggestion
**timer()** creates a stream that emits a value every time the interval (in milliseconds) we provide has passed, essentially working like `setInterval`, but for streams.
```

Reviewer: `timer` is like `setTimeout` followed by `setInterval` when it's passed 2 arguments since it adds an initial delay as specified by the first argument. `interval` is like `setInterval`.

How do we express this?

Author: I don't know, it wa written in a sloppy way in the original, so I just removed it :D

Reviewer: I'll leave a thread about this in [our Part 3 discussion](https://github.com/this-is-learning/rxjs-fundamentals-course/discussions/3).

```ts
import { timer } from 'rxjs';

// First value will be emitted immediately, the next one in 1000ms

const sourceFirst = timer(0, 1000);

// Will log:

//. 0 immediately

//. 1 in 1.0 second

sourceFirst.subscribe(value => console.log(value));

// First value in 0.5 second, next one in 1 second

const sourceSecond = timer(500, 1000);

// Will log:

//. 0 in 0.5 sec

//. 1 in 1.5 sec

sourceSecond.subscribe(value => console.log(value));
```

## **interval**
Next, let's learn how to create a stream that emits over time, but repeatedly (like `setInterval`). For that purpose, we have the **interval** function:

```ts
import { interval } from 'rxjs';

const source$ = interval(1000);

// Will log:

//. 0 in 1 second

//. 1 in 2 seconds

//. 2 in 3 seconds

// and so on

source$.subscribe(value => console.log(value))
```

## **combineLatest()**

Now let's combine values of those streams! We will learn to use the **combineLatest()** operator.

> **combineLatest()** unites the values of all the provided streams, end emits the latest emitted values of each one of them in an `Array`. So, for example, if one stream emits, that value will be combined with all the latest values emitted from other `Observables` in that array, and emitted together.

Reviewer: ```suggestion
**combineLatest()** unites the values of all the provided streams and emits the latest emitted values of each one of them in an `Array`. So, for example, if one stream emits, that value will be combined with all the latest values emitted from other `Observables` in that array, and emitted together.
```

Reviewer: Not expressed very clearly. Consider rephrasing.

> **Important!** Notice that **combineLatest** will only start emitting after each source `Observable` has emitted at least once. Previous incomplete values will be ignored 

Reviewer: > Previous incomplete values will be ignored 

What does this mean?

Author: For example, if we have combineLatest([interval(500), interval(1000)]), the first emission by interval(500) will be ignored

Reviewer: The *incomplete values* term is what's confusing me.

How about:
```suggestion
**Important!** Notice that **combineLatest** only starts emitting after each source `Observable` has emitted at least one value. Values emitted before this moment are ignored by the resulting `Observable`.
```

```ts
import { interval, combineLatest } from 'rxjs';

const sourceFirst = interval(500);

const sourceSecond = interval(1000);

const source = combineLatest([sourceFirst, sourceSecond]);

// Will log:

//. [1, 0] in 1 second

//. [2, 0] in 1.5 seconds

// [3, 1] in 2 seconds

// and so on

source.subscribe(value => console.log(value));
```

> Despite the values of two streams are generated at different times, we managed to combine them into a single value, which allows us to handle values from two different streams simultaneously. **combineLatest** allows us to combine as many `Observables` as we want, not just two. 

Reviewer: ```suggestion
Despite the values of these two streams are generated at different times, we manage to combine them into a single value which allows us to handle values from two different streams simultaneously. **combineLatest** allows us to combine as many `Observables` as we want, not just two.
```

> ## **withLatestFrom()**

Reviewer: Nice catch! 😄

As you could see, we received a new value each time any of the two streams emitted. But what if we need to receive a value only when the source `Observable` emits? Let's achieve that using the **withLatestFrom()** operator.

**withLatestFrom()** combines the source `Observable` with another one, and emits a combined value only when the source one emits.

```ts
import { interval } from 'rxjs';

import { withLatestFrom } from 'rxjs/operators';

const sourceFirst = interval(1000);

const sourceSecond = interval(500);

> const source = sourceFirst.pipe(withLatestFrom(sourceSecond));

Reviewer 2: Nice to have: Examples with some additional log information. Something like this -

```
const source = sourceFirst.pipe(
  withLatestFrom(sourceSecond),
  map(([first, second]) => {
    return `First Source : ${first} Second Source : ${second}`;
  })
);
```

Reviewer: Great suggestion, @prawin-s. This PR is mostly about translating the content from Russian to English. You're very welcome to add this suggestion to [the Part 3 discussion](https://github.com/this-is-learning/rxjs-fundamentals-course/discussions/3) or submit another PR 👐

Author: I agree. That's why I primarily wanted te examples to be reviewed

Reviewer: I'll add a thread about code snippet review to [our Part 3 discussion](https://github.com/this-is-learning/rxjs-fundamentals-course/discussions/3) and link to @prawin-s' comment.

// Will log:

//. [0, 1] in 1.0 second

// [1, 3] in 2.0 seconds

// [2, 5] in 3.0 seconds

source.subscribe(value => console.log(value));
```

> The only difference from **combineLatest** is that it will only emit when the source emits. But it won't emit if the other `Observable` has not emitted at all, even if the source emits, just like **combineLatest**. 

Reviewer: >But it won't emit if the other `Observable` has not emitted at all, even if the source emits, just like **combineLatest**. 

Consider rephrasing. Not very clear.

Author: I changed it a bit to be more clear. Do you think it is better now?

#

# **Handling errors in streams**

In real life application some exceptional situations may arise, in which cases we will need to correctly handle errors. Let's create an error and see how it goes.

```ts
import { from } from 'rxjs';

import { map } from 'rxjs/operators';

const source = from([1, 2, 3]).pipe(
  // Something went wrong!

  map(() => {

>     throw newError("Unexpected 🙀!");

Reviewer: ```suggestion
    throw new Error("Unexpected 🙀!");
```

  })
);

// An Error will be thrown:

//. Error: Unexpected 🙀!

source.subscribe(value => console.log(value));
```

> In this case, this `Error` is not being handled at all.To handle it, we will need o learn how to use the **catchError()** operator.

Reviewer: ```suggestion
In this case, this `Error` is not being handled at all. To handle it, we have to learn how to use the **catchError()** operator.
```

> **catchError()** handles all the errors that happen inside the stream. Notice that when an error happens, the old stream completes, so we need to return a new stream from that operator.

Reviewer: ```suggestion
**catchError()** handles all errors that occur inside of a stream. Notice that when an error occurs, the old stream completes, so we need to return a new stream from the `catchError` operator.
```

Reviewer: Consider mentioning what happens if we don't return an observable from `catchError`.

```ts
import { from, of } from 'rxjs';

import { map, catchError } from 'rxjs/operators';

const source = from([1, 2, 3]).pipe(
  // Something went wrong!

  map(() => {

>     throw newError("Unexpected 🙀!");

Reviewer 2: Import throwError from rxjs and use that instead of newError.

Author: I think you are right, but this section does not mention `throwError` at all. I guess if we do import it, we have to write a section explaining how it works

Author: @LayZeeDK what do you think?

Reviewer: I think we should check whether throwError is explained in other parts. If not, discuss where it would best fit.

Author: I think, if it is not provided elsewhere, we should just go ahead and explain it here, where we mention error handling

Reviewer: I just went through the content. We use `throwError` one time, in Part 5. `catchError` is used one time in Part 5 and mentioned in a comment in Part 4.

It's a good suggestion to add it here. If we later see it fitting better in another part, we can just move it.

But let's add this section in a separate PR so that this one can remain about the translation 🙂

Reviewer: I'll add a thread to [our Part 3 discussion](https://github.com/this-is-learning/rxjs-fundamentals-course/discussions/3).

>     throw newError("Unexpected 🙀!");

Reviewer: ```suggestion
    throw new Error("Unexpected 🙀!");
```

  }),

  // Let's handle it!

  catchError(() => of("Expected 😻!")),
);

// Will log:

//. Expected 😻!

source.subscribe((value => console.log(value));
```

> Now w can be sure the code will work correctly, and all the errors will be handled.

Reviewer: ```suggestion
Now we can rest assured that the code will work correctly and all errors are handled.
```

#

# **In Conclusion**

> We got to know the most common **RxJS** operators and learned to: change values in a stream, filter values, combine streams and handle errors.

Reviewer: ```suggestion
We got to know the most common **RxJS** operators and learnt to:
- Change values in a stream
- Filter values
- Combine streams
- Handle errors
```

Reviewer: Nice summary 🙌

// Will log:

//. [1, 0] in 1 second

//. [2, 0] in 1.5 seconds

// [3, 1] in 2 seconds

// and so on

source.subscribe(value => console.log(value));
```

Despite the values of these two streams are generated at different times, we manage to combine them into a single value which allows us to handle values from two different streams simultaneously. **combineLatest** allows us to combine as many `Observables` as we want, not just two.

## **withLatestFrom()**

As you could see, we received a new value each time any of the two streams emitted. But what if we need to receive a value only when the source `Observable` emits? Let's achieve that using the **withLatestFrom()** operator.

**withLatestFrom()** combines the source `Observable` with another one, and emits a combined value only when the source one emits.

```ts
import { interval } from 'rxjs';

import { withLatestFrom } from 'rxjs/operators';

const sourceFirst = interval(1000);

const sourceSecond = interval(500);

const source = sourceFirst.pipe(withLatestFrom(sourceSecond));

// Will log:

//. [0, 1] in 1.0 second

// [1, 3] in 2.0 seconds

// [2, 5] in 3.0 seconds

source.subscribe(value => console.log(value));
```

> The only difference from **combineLatest** is that it will only emit when the source `Observable` emits. But it won't emit if the `Observable` we proided to **withLatestFrom** has not emitted at all, even if the source `Observable` emits, just like **combineLatest**, so for instance, in this example, the first emission from `sourceSecond` will be ignored. 

Reviewer: Does this sound correct, @Armenvardanyan95?
```suggestion
The only difference from **combineLatest** is that it will only emit when the source `Observable` emits. It won't emit if the `Observable` we passed to **withLatestFrom** has not emitted at all, even when the source `Observable` emits. This is similar to **combineLatest**, for instance in this example, the first emission from `sourceSecond` will be ignored. 
```
